app/android/src/uk/co/lutraconsulting/CameraActivity.java

```diff
@@ -73,85 +100,240 @@ protected void onCreate(Bundle savedInstanceState) {
             }
         }
 
-        //throw new IOException("Tralalal!#@#$@JAVA");
-
         return;
     }
 
+    /**
+     * Listeners for the sensors are registered in this callback so that
+     * they can be unregistered in onStop().
+     */
+    @Override
+    protected void onStart() {
+        super.onStart();
+
+        if (mSensorAccelerometer != null) {
+            mSensorManager.registerListener(this, mSensorAccelerometer,
+                    SENSOR_DELAY);
+        }
+        if (mSensorMagnetometer != null) {
+            mSensorManager.registerListener(this, mSensorMagnetometer,
+                    SENSOR_DELAY);
+        }
+    }
+
+    @Override
+    public void onSensorChanged(SensorEvent sensorEvent) {
+        // The sensor type (as defined in the Sensor class).
+        int sensorType = sensorEvent.sensor.getType();
+
+        // The sensorEvent object is reused across calls to onSensorChanged().
+        // clone() gets a copy so the data doesn't change out from under us
+        switch (sensorType) {
+            case Sensor.TYPE_ACCELEROMETER:
+                mAccelerometerData = sensorEvent.values.clone();
+                break;
+            case Sensor.TYPE_MAGNETIC_FIELD:
+                mMagnetometerData = sensorEvent.values.clone();
+                break;
+            default:
+                return;
+        }
+        // Compute the rotation matrix: merges and translates the data
+        // from the accelerometer and magnetometer, in the device coordinate
+        // system, into a matrix in the world's coordinate system.
+        //
+        // The second argument is an inclination matrix, which isn't
+        // used in this example.
+        float[] rotationMatrix = new float[9];
+        boolean rotationOK = SensorManager.getRotationMatrix(rotationMatrix,
+                null, mAccelerometerData, mMagnetometerData);
+
+        // Remap the matrix based on current device/activity rotation.
+        float[] rotationMatrixAdjusted = new float[9];
+        rotationMatrixAdjusted = rotationMatrix.clone();
+
+        // Get the orientation of the device (azimuth, pitch, roll) based
+        // on the rotation matrix. Output units are radians.
+        float orientationValues[] = new float[3];
+        if (rotationOK) {
```

Comment:
What if rotation is not OK? can it happen, and if yes, shouldn't we just skip everything else to make sure we don't use uninitialized values?